### PR TITLE
Complete the TiffIFD/TiffEntry interface cleanup

### PR DESCRIPTION
**Pull request by axxel**
* make getter function names consistent
* complete const-correctness of TiffIFD interface

### CODE REVIEW

File: src/librawspeed/decoders/AbstractTiffDecoder.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+#include "decoders/RawDecoderException.h" // for ThrowRDE
+
+namespace RawSpeed {
+
+const TiffIFD* AbstractTiffDecoder::getIFDWithLargestImage(TiffTag filter) const
+{
+  std::vector<const TiffIFD*> ifds = mRootIFD->getIFDsWithTag(filter);
+
+  if (ifds.empty())
+    ThrowRDE("AbstractTiffDecoder: No suitable IFD with IMAGEWIDTH tag found.");
```
Comment (LebedevRI) [inline src/librawspeed/decoders/AbstractTiffDecoder.cpp]: with `filter` you mean

Comment (axxel) [reply src/librawspeed/decoders/AbstractTiffDecoder.cpp]: absolutely right. see update.

**Comment by axxel** [comment]
@LebedevRI Please merge (as early as convenient :)).

File: src/librawspeed/decoders/MosDecoder.cpp [hunk 2]
```diff
@@ -124,17 +123,15 @@ RawImage MosDecoder::decodeRawInternal() {
...
+    raw = mRootIFD->getIFDWithTag(TILEOFFSETS);
     off = raw->getEntry(TILEOFFSETS)->getU32();
-  } else {
-    data = mRootIFD->getIFDsWithTag(CFAPATTERN);
-    if (!data.empty()) {
-      raw = data[0];
-      off = raw->getEntry(STRIPOFFSETS)->getU32();
-    } else
-      ThrowRDE("MOS Decoder: No image data found");
+  } catch(...) {
```
Comment (LebedevRI) [inline src/librawspeed/decoders/MosDecoder.cpp]: using exceptions for normal codepath handling, wtf c++?

File: src/librawspeed/decoders/Rw2Decoder.cpp [hunk 3]
```diff
@@ -255,19 +253,14 @@ void Rw2Decoder::decodeMetaDataInternal(CameraMetaData *meta) {
...
-  // bool isOldPanasonic = false;
+  const TiffIFD* raw;
 
-  if (data.empty()) {
-    if (!mRootIFD->hasEntryRecursive(STRIPOFFSETS))
-      ThrowRDE("RW2 Decoder: No image data found");
-    // isOldPanasonic = true;
-    data = mRootIFD->getIFDsWithTag(STRIPOFFSETS);
+  try {
+    raw = mRootIFD->getIFDWithTag(PANASONIC_STRIPOFFSET);
```
Comment (LebedevRI) [inline src/librawspeed/decoders/Rw2Decoder.cpp]: i really don't like this pattern

**Review by LebedevRI — CHANGES_REQUESTED** [review]
i *really* don't like that pattern of using exception handling as general-purpose codeflow mechaninsm that you are adding e.g. in `MosDecoder.cpp`.

**Comment by axxel** [comment]
Well, there is _tons_ of 'prior art' in the code base for using exceptions to deal with 'non fatal' error conditions and it is not time critical either. So I don't see any reason why this would be an issue. But I see even less of a point in starting an argument over this detail. See appended patch.

**Comment by axxel** [comment]
@LebedevRI you probably noticed, but just in case you did not and are waiting for this to be fixed: the compile issue seems to be caused by your recent gtest-option commit or something else build system related.

**Comment by LebedevRI** [comment]
@axxel if the build breaks, i have all kinds of bells and whistles that start start signalling, so i do notice :)

**Comment by axxel** [comment]
> so i do notice :)

:). Your usual approach to have a 'next' pr to get ci-feedback before merging seemed like a valid strategy to prevent breaking the 'default' build.

**Comment by LebedevRI** [comment]
Just re-triggering the build does not seem to help, so please rebase.

**Comment by LebedevRI** [comment]
> Well, there is tons of 'prior art' in the code base for using exceptions to deal with 'non fatal' error conditions and it is not time critical either. So I don't see any reason why this would be an issue. 

Since you want to touch everything immediately, you leave me no other option on how to change 'prior art' than to complain about it when i see it in the diff

File: src/librawspeed/decoders/Rw2Decoder.cpp (hunk 3 above)
Comment (LebedevRI) [inline src/librawspeed/decoders/Rw2Decoder.cpp]: this is still here

File: src/librawspeed/decoders/Rw2Decoder.cpp [hunk 4]
```diff
@@ -255,24 +251,19 @@ void Rw2Decoder::decodeMetaDataInternal(CameraMetaData *meta) {
...
 
-  // bool isOldPanasonic = false;
-
-  if (data.empty()) {
-    if (!mRootIFD->hasEntryRecursive(STRIPOFFSETS))
-      ThrowRDE("RW2 Decoder: No image data found");
-    // isOldPanasonic = true;
-    data = mRootIFD->getIFDsWithTag(STRIPOFFSETS);
+  try {
+    raw = mRootIFD->getIFDWithTag(PANASONIC_STRIPOFFSET);
```
Comment (LebedevRI) [inline src/librawspeed/decoders/Rw2Decoder.cpp]: this is still here

Comment (axxel) [reply src/librawspeed/decoders/Rw2Decoder.cpp]: indeed

**Review by LebedevRI — APPROVED** [review]
Okay, seems to work :)
Will merge.

**Comment by LebedevRI** [comment]
@axxel thank you.

**Pull request merged**